commons/src/main/java/io/aiven/kafka/tieredstorage/commons/storage/s3/S3StorageConfig.java

```diff
@@ -135,7 +135,10 @@ AmazonS3 s3Client() {
         } else {
             final AwsClientBuilder.EndpointConfiguration endpointConfiguration =
                 new AwsClientBuilder.EndpointConfiguration(s3ServiceEndpoint, region);
-            s3ClientBuilder.withEndpointConfiguration(endpointConfiguration);
+            // TODO: fix minIO local environment to support virtual host or enable path style via config
+            // https://stackoverflow.com/questions/59693471/how-to-setup-minio-server-to-use-virtual-hosted-style
+            s3ClientBuilder.withEndpointConfiguration(endpointConfiguration)
+                .withPathStyleAccessEnabled(true);
```

Comment:
So, what's the problem with Minio once again? I run
```shell
docker run -p 9000:9000 -p 9001:9001   quay.io/minio/minio server /data --console-address ":9001"
```

and then this worked for me:
```java
AWSCredentials credentials = new BasicAWSCredentials("minioadmin", "minioadmin");
AwsClientBuilder.EndpointConfiguration endpoint = new AwsClientBuilder.EndpointConfiguration("http://172.17.0.2:9000", "x");
AmazonS3 s3 = AmazonS3ClientBuilder.standard()
    .withCredentials(new AWSStaticCredentialsProvider(credentials))
    .withEndpointConfiguration(endpoint)
    .build();
s3.listObjectsV2("xxx").getObjectSummaries().forEach(System.out::println);
```

Reply:
I guess it's a combination of minio and docker compose. When uploading objects with minio configuration, the following host is used `<bucket_name>.minio`, and requests are failing with:

```
[2023-05-22 15:14:01,156] ERROR [RemoteLogManager=0 partition=TopicIdPartition{topicId=zRimZy3gQf-VBdjdqPdgMA, topicPartition=t1-3}] Error occurred while copying log segments of partition: TopicIdPartition{topicId=zRimZy3gQf-VBdjdqPdgMA, topicPartition=t1-3} (kafka.log.remote.RemoteLogManager$RLMTask)
com.amazonaws.SdkClientException: Unable to execute HTTP request: kafka-ts-us-east-1.minio
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleRetryableException(AmazonHttpClient.java:1219)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1165)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:814)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:781)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:755)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:715)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:697)
        at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:561)
        at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:541)
        at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5456)
        at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5403)
        at com.amazonaws.services.s3.AmazonS3Client.initiateMultipartUpload(AmazonS3Client.java:3727)
        at io.aiven.kafka.tieredstorage.commons.storage.s3.S3OutputStream.initiate(S3OutputStream.java:68)
        at io.aiven.kafka.tieredstorage.commons.storage.s3.S3OutputStream.write(S3OutputStream.java:87)
        at java.base/java.io.InputStream.transferTo(Unknown Source)
        at io.aiven.kafka.tieredstorage.commons.storage.s3.S3Storage.upload(S3Storage.java:50)
        at io.aiven.kafka.tieredstorage.commons.UniversalRemoteStorageManager.copyLogSegmentData(UniversalRemoteStorageManager.java:190)
        at kafka.log.remote.ClassLoaderAwareRemoteStorageManager.$anonfun$copyLogSegmentData$1(ClassLoaderAwareRemoteStorageManager.scala:54)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at kafka.log.remote.ClassLoaderAwareRemoteStorageManager.withClassLoader(ClassLoaderAwareRemoteStorageManager.scala:34)
        at kafka.log.remote.ClassLoaderAwareRemoteStorageManager.copyLogSegmentData(ClassLoaderAwareRemoteStorageManager.scala:54)
        at kafka.log.remote.RemoteLogManager$RLMTask.$anonfun$copyLogSegmentsToRemote$8(RemoteLogManager.scala:450)
        at kafka.log.remote.RemoteLogManager$RLMTask.$anonfun$copyLogSegmentsToRemote$8$adapted(RemoteLogManager.scala:418)
        at scala.collection.immutable.Stream.foreach(Stream.scala:533)
        at kafka.log.remote.RemoteLogManager$RLMTask.$anonfun$copyLogSegmentsToRemote$2(RemoteLogManager.scala:418)
        at kafka.log.remote.RemoteLogManager$RLMTask.$anonfun$copyLogSegmentsToRemote$2$adapted(RemoteLogManager.scala:394)
        at scala.Option.foreach(Option.scala:407)
        at kafka.log.remote.RemoteLogManager$RLMTask.copyLogSegmentsToRemote(RemoteLogManager.scala:394)
        at kafka.log.remote.RemoteLogManager$RLMTask.run(RemoteLogManager.scala:564)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.base/java.util.concurrent.FutureTask.runAndReset(Unknown Source)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.net.UnknownHostException: kafka-ts-us-east-1.minio
        at java.base/java.net.InetAddress$CachedAddresses.get(Unknown Source)
        at java.base/java.net.InetAddress.getAllByName0(Unknown Source)
        at java.base/java.net.InetAddress.getAllByName(Unknown Source)
        at java.base/java.net.InetAddress.getAllByName(Unknown Source)
        at com.amazonaws.SystemDefaultDnsResolver.resolve(SystemDefaultDnsResolver.java:27)
        at com.amazonaws.http.DelegatingDnsResolver.resolve(DelegatingDnsResolver.java:38)
        at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:112)
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:376)
        at jdk.internal.reflect.GeneratedMethodAccessor14.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.base/java.lang.reflect.Method.invoke(Unknown Source)
        at com.amazonaws.http.conn.ClientConnectionManagerFactory$Handler.invoke(ClientConnectionManagerFactory.java:76)
        at com.amazonaws.http.conn.$Proxy7.connect(Unknown Source)
        at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:393)
        at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
        at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
        at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:56)
        at com.amazonaws.http.apache.client.impl.SdkHttpClient.execute(SdkHttpClient.java:72)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1346)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1157)
        ... 33 more
```

I tried the suggested code, and it doesn't fail, but it doesn't return results neither. e.g. if you try to get an object, you would get something like:
```
Exception in thread "main" com.amazonaws.services.s3.model.AmazonS3Exception: The specified bucket does not exist (Service: Amazon S3; Status Code: 404; Error Code: NoSuchBucket; Request ID: 176180918592C669; S3 Extended Request ID: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855; Proxy: null), S3 Extended Request ID: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

but if s3 client has path style access enabled, it should work:

```
        AmazonS3 s3 = AmazonS3ClientBuilder.standard()
            .withCredentials(new AWSStaticCredentialsProvider(credentials))
            .withEndpointConfiguration(endpoint)
+            .withPathStyleAccessEnabled(true)
            .build();

```

Reply:
Could you share your compose file?

Reply:
https://github.com/aiven/tiered-storage-for-apache-kafka/blob/new-implementation/docker/s3-minio/compose.yml#L71-L80

Reply:
I see. Yeah, it seems a DNS issue. There may be options like reverse proxy or DNS configuring (e.g. add
```
    networks:
      default:
        aliases:
          - bucket_name.minio
```
to the `minio` container config), using the minio IP address instead of the DNS name. And of course the path-style access.
Let's make it configurable, though: having hardcoded path-style is not great.